crates/stages/src/pipeline/set.rs

```diff
@@ -0,0 +1,207 @@
+        self.order.push(stage.id());
+        self.upsert_stage_state(Box::new(stage), target_index);
+        self
+    }
+
+    /// Adds the given [`StageSet`] to the end of this set.
+    ///
+    /// If a stage is in both sets, it is removed from its previous place in this set. Because of
+    /// this, it is advisable to merge sets first and re-order stages after if needed.
+    pub fn add_set<Set: StageSet<DB>>(mut self, set: Set) -> Self {
+        for stage in set.build().finish() {
+            let target_index = self.order.len();
+            self.order.push(stage.id());
+            self.upsert_stage_state(stage, target_index);
+        }
+        self
+    }
+
+    /// Adds the given [`Stage`] before the stage with the given [`StageId`].
+    ///
+    /// If the stage was already in the group, it is removed from its previous place.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the dependency stage is not in this set.
+    pub fn add_before<S: Stage<DB> + 'static>(mut self, stage: S, before: StageId) -> Self {
+        let target_index = self.index_of(before);
+        self.order.insert(target_index, stage.id());
+        self.upsert_stage_state(Box::new(stage), target_index);
+        self
+    }
+
+    /// Adds the given [`Stage`] after the stage with the given [`StageId`].
+    ///
+    /// If the stage was already in the group, it is removed from its previous place.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the dependency stage is not in this set.
+    pub fn add_after<S: Stage<DB> + 'static>(mut self, stage: S, after: StageId) -> Self {
+        let target_index = self.index_of(after) + 1;
+        self.order.insert(target_index, stage.id());
+        self.upsert_stage_state(Box::new(stage), target_index);
+        self
+    }
+
+    /// Enables the given stage.
+    ///
+    /// All stages within a [`StageSet`] are enabled by default.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the stage is not in this set.
+    pub fn enable(mut self, stage_id: StageId) -> Self {
+        let mut entry =
+            self.stages.get_mut(&stage_id).expect("Cannot enable a stage that is not in the set.");
+        entry.enabled = true;
+        self
+    }
+
+    /// Disables the given stage.
+    ///
+    /// The disabled [`Stage`] keeps its place in the set, so it can be used for ordering with
+    /// [`add_before`] or [`add_after`], or it can be re-enabled.
+    ///
+    /// All stages within a [`StageSet`] are enabled by default.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the stage is not in this set.
+    pub fn disable(mut self, stage_id: StageId) -> Self {
+        let mut entry =
+            self.stages.get_mut(&stage_id).expect("Cannot disable a stage that is not in the set.");
+        entry.enabled = false;
+        self
+    }
+
+    /// Consumes the builder and returns the contained [`Stage`]s in the order specified.
+    pub fn finish(mut self) -> Vec<Box<dyn Stage<DB>>> {
```

Comment:
For consistency?

```suggestion
    pub fn build(mut self) -> Vec<Box<dyn Stage<DB>>> {
```

Reply:
It just clashes with `build` in the `StageSet` trait. Do you have suggestions for the naming there?